applications/nrf5340_audio/src/audio_module/src/amod_api.h

```diff
@@ -0,0 +1,492 @@
+	 *
+	 * @param handle		 A handle to this module instance
+	 * @param configuration  A pointer to the modules configuration to set
+	 *
+	 * @return 0 if successful, error value
+	 */
+	int (*configuration_set)(struct handle *handle, struct amod_configuration *configuration);
+
+	/**
+	 * @brief  Function to get the configuration of a module.
+	 *
+	 * @param handle		 A handle to this module instance
+	 * @param configuration  A pointer to the modules current configuration
+	 *
+	 * @return 0 if successful, error value
+	 */
+	int (*configuration_get)(struct handle *handle, struct amod_configuration *configuration);
+
+	/**
+	 * @brief Start a module processing data.
+	 *
+	 * @param handle  The handle for the module to start
+	 *
+	 * @return 0 if successful, error value
+	 */
+	int (*start)(struct handle *handle);
+
+	/**
+	 * @brief Stop a module processing data.
+	 *
+	 * @param handle  The handle for the module to stop
+	 *
+	 * @return 0 if successful, error value
+	 */
+	int (*stop)(struct handle *handle);
+
+	/**
+	 * @brief The core data processing function for the module. Can be either an
+	 *		input or output or processor module type.
+	 *
+	 * @param handle	 A handle to this module instance
+	 * @param handle_tx  Pointer to the input audio block or NULL for an input module
+	 * @param block_rx   Pointer to the output audio block or NULL for an output module
+	 *
+	 * @return 0 if successful, error value
+	 */
+	int (*data_process)(struct handle *handle, struct aobj_block *handle_tx,
+			    struct aobj_block *block_rx);
+};
+
+/**
+ * @brief A modules minimum description.
+ */
+struct amod_description {
+	/* The module base name */
+	char *name;
+
+	/* The module type */
+	enum amod_type type;
+
+	/* A pointer to the functions to implment the module */
+	const struct amod_functions *functions;
+};
+
+/**
+ * @brief Module's thread configuration structure.
+ */
+struct amod_thread_configuration {
+	/* Thread stack */
+	k_thread_stack_t *stack;
+
+	/* Thread stack size */
+	size_t stack_size;
+
+	/* Thread priority */
+	int priority;
+
+	/* Pointer to a modules data receiver fifo */
+	struct data_fifo *msg_rx;
+
+	/* Pointer to a modules data transmitter fifo */
+	struct data_fifo *msg_tx;
+
+	/* Pointer to the data buffer slab */
+	struct k_mem_slab *data_slab;
+
+	/* Size of a data block taken from the slab */
+	size_t data_size;
+};
+
+/**
+ * @brief Module's generic set-up structure.
+ */
+struct amod_parameters {
+	/* The modules private description */
+	struct amod_description *description;
+
+	/* The modules thread setting */
+	struct amod_thread_configuration thread;
+};
+
+/**
+ * @brief Private module handle.
+ */
+struct amod_handle {
+	/* Unique name of this module instance */
+	char name[CONFIG_AMOD_NAME_SIZE];
+
+	/* The modules description */
+	struct amod_description *description;
+
+	/* Current state of the module */
+	enum amod_state state;
+
+	/* Previous state of the module */
+	enum amod_state previous_state;
```

Comment:
Why do we want/need to store the previous state?

Reply:
It can be useful when transitioning to the running state. For example we may wish to move differently from stooped state to running, compared with moving from configured state to running.